module/zfs/arc.c

```diff
@@ -8584,15 +8952,67 @@ l2arc_add_vdev(spa_t *spa, vdev_t *vd)
 	list_create(&adddev->l2ad_buflist, sizeof (arc_buf_hdr_t),
 	    offsetof(arc_buf_hdr_t, b_l2hdr.b_l2node));
 
+	list_create(&adddev->l2ad_lbptr_list, sizeof (l2arc_lb_ptr_buf_t),
+	    offsetof(l2arc_lb_ptr_buf_t, node));
+
 	vdev_space_update(vd, 0, 0, adddev->l2ad_end - adddev->l2ad_hand);
 	zfs_refcount_create(&adddev->l2ad_alloc);
+	zfs_refcount_create(&adddev->l2ad_log_blk_count);
 
 	/*
-	 * Add device to global list
+	 * The L2ARC has to hold at least the payload of one log block for
+	 * them to be restored (persistent L2ARC). The payload of a log block
+	 * depends on the amount of its log entries. We always write log blocks
+	 * with 1023 entries. How many of them are committed or restored depends
+	 * on the size of the L2ARC device. Thus the maximum payload of
+	 * one log block is 1023 * SPA_MAXBLOCKSIZE = 16GB. If the L2ARC device
+	 * is less than that, we reduce the amount of committed and restored
+	 * log entries per block so as to enable persistence.
+	 */
+	if (adddev->l2ad_end < l2arc_rebuild_blocks_min_l2size) {
+		adddev->l2ad_log_entries = 0;
+	} else {
+		adddev->l2ad_log_entries = MIN((adddev->l2ad_end -
+		    adddev->l2ad_start) >> SPA_MAXBLOCKSHIFT,
+		    L2ARC_LOG_BLK_MAX_ENTRIES);
+	}
+
+	/*
+	 * Read the device header, and if hdr->dh_log_blk_ent is not equal to
+	 * the calculated one do not rebuild L2ARC.
 	 */
 	mutex_enter(&l2arc_dev_mtx);
+	if ((err = l2arc_dev_hdr_read(adddev)) != 0) {
+		/* device header corrupted, start a new one */
+		bzero(hdr, adddev->l2ad_dev_hdr_asize);
+		rebuild = B_FALSE;
+	}
+
+	/*
+	 * Add device to global list
+	 */
 	list_insert_head(l2arc_dev_list, adddev);
 	atomic_inc_64(&l2arc_ndev);
+	if (rebuild && l2arc_rebuild_enabled && hdr->dh_log_blk_ent > 0 &&
+	    hdr->dh_log_blk_count > 0) {
+		/*
+		 * Just mark the device as pending for a rebuild. We won't
+		 * be starting a rebuild in line here as it would block pool
+		 * import. Instead spa_load_impl will hand that off to an
+		 * async task which will call l2arc_spa_rebuild_start.
+		 */
+		adddev->l2ad_rebuild = B_TRUE;
+	} else if (err != 0 && spa_writeable(spa)) {
+		/*
+		 * If we are not rebuilding the L2ARC and if the header was
+		 * corrupted or non-existent, create a new one.
+		 * We cannot update the header of the device here if the L2ARC
+		 * is going to be rebuilt since dh_log_blk_count will be set to
+		 * 0.
+		 */
+		l2arc_dev_hdr_update(adddev);
```

Comment:
Do we still write the l2arc hdr if `adddev->l2ad_log_entries` is 0?

Reply:
Yes, we do. It is only 512 bytes, so I left it that way.